### PR TITLE
fix(daft-file): release GIL in PyDaftFile methods to prevent deadlock

### LINKED ISSUE #7256: GIL deadlock in PyDaftFile methods when reading S3 files under concurrent scan tasks
### Describe the bug

In Ray distributed mode, `PyDaftFile`'s `#[pymethods]` (`_from_file_reference`, `read`, `seek`, `_supports_range_requests`, `__enter__`) hold the Python GIL while calling `block_within_async_context`, which blocks via `rx.recv()` waiting for async S3 I/O on DAFTIO tokio worker threads. When concurrent scan tasks need the GIL on the same DAFTIO threads (via `Python::attach`), all threads deadlock permanently — no error, no timeout, no recovery.

Confirmed with py-spy (BackgroundEventLoop thread `active+gil` stuck at `File.open()`), kernel thread stacks (all 8 DAFTIO threads on `futex_wait`), and `ss` (0 TCP connections to S3 — the request was never sent).

The `exists()` method in the same file (`python.rs:102`) correctly uses `py.detach()`. The 5 affected methods missed this pattern.

### To Reproduce

**Environment**: Ray distributed mode, `os.cpu_count()` >= 8, S3-compatible storage with latency >= 50ms.

```python
import daft

daft.set_runner_ray()
daft.set_planning_config(default_io_config=daft.io.IOConfig(
    s3=daft.io.S3Config(
        endpoint_url="https://your-s3-compatible-endpoint",
        region_name="your-region",
        key_id="...",
        access_key="...",
    )
))

df = daft.read_video_frames("s3://bucket/videos/", image_height=640, image_width=640)
df.count_rows()  # hangs indefinitely after processing ~19K frames
```

**All three conditions required to trigger:**
1. Multiple concurrent `run_plan` tasks per worker (Ray distributed mode)
2. S3 latency >= ~50ms (cross-region S3, MinIO, GCS, Azure Blob, Alibaba Cloud OSS, etc.)
3. `os.cpu_count()` >= 8 on the worker machine (DAFTIO threads = `min(8, cpu_count)`)

AWS g6.xlarge (4 vCPU, same-AZ S3 < 1ms) does not trigger due to low thread count and latency.

### Expected behavior

The pipeline should complete without deadlocking, regardless of S3 latency or DAFTIO thread count.

### Component(s)

Cloud Storage (S3/GCS/Azure), Data Sources & Sinks, Distributed Runner (flotilla)

### Additional context

**Deadlock cycle:**
```
① BackgroundEventLoop: holds GIL → File.open() → rx.recv() blocks (no py.detach())
② 5 DAFTIO threads: queue.get() waiting for data from ①
③ 3 DAFTIO threads: Python::attach waiting for GIL held by ①
④ S3 task: needs a DAFTIO thread → all 8 blocked (②+③) → never executes
→ rx.recv() never returns → GIL never released → permanent deadlock
```

**Diagnostic evidence:**
- `py-spy dump`: BackgroundEventLoop `(active+gil)` at `File.open()`, 5 threads at `queue.get()`, 3 threads `(active)` no Python stack (waiting for GIL in Rust `Python::attach`)
- Kernel stacks: `DAFTIO-0` through `DAFTIO-7` all on `futex_wait` for 24+ hours
- `ss -tnopi | grep 443`: 0 connections to S3

**Affected code** (`src/daft-file/src/python.rs`):
- `PyFileReference::__enter__` (line 33)
- `PyDaftFile::_from_file_reference` (line 150)
- `PyDaftFile::read` (line 155)
- `PyDaftFile::seek` (line 193)
- `PyDaftFile::_supports_range_requests` (line 260)

**Proposed fix**: Wrap blocking calls in `py.detach()`, same pattern as `exists()` at line 102. For `read()`/`seek()`, use `cursor.take()` + restore to satisfy Rust ownership.

**Verified**: 24+ hour deadlock → 9.6 minutes after fix.

### PR DESCRIPTION
**Pull request by zhouwenjiajia**
PyDaftFile's #[pymethods] (_from_file_reference, read, seek, _supports_range_requests) and PyFileReference::__enter__ hold the Python GIL while calling block_within_async_context, which blocks via rx.recv() waiting for async S3 I/O on DAFTIO tokio threads. When concurrent scan tasks need the GIL on the same DAFTIO threads (via Python::attach), all threads deadlock — the GIL holder waits for a DAFTIO thread to execute the S3 request, while all DAFTIO threads wait for the GIL.

The fix wraps the blocking I/O calls in py.detach() to release the GIL before blocking, following the existing pattern used by the exists() method in the same file. For read() and seek(), cursor ownership is temporarily moved into the detach closure via take() and restored after completion.

Deadlock conditions: Ray distributed mode + multiple concurrent scan tasks per worker + S3 latency >= ~50ms + DAFTIO threads >= 8 (os.cpu_count() >= 8). The official AWS benchmark (4 vCPU, S3 <1ms) does not trigger this due to low thread count and latency.

## Changes Made

Modified `src/daft-file/src/python.rs` — added `py.detach()` to 5 methods that
hold the GIL while calling `block_within_async_context`:

- `PyFileReference::__enter__`: added `py: Python<'_>`, wrapped `load_blocking()`
  in `py.detach()` with cloned `FileReference`.
- `PyDaftFile::_from_file_reference`: same pattern — clone data, `py.detach()`,
  call `load_blocking()` inside.
- `PyDaftFile::read`: added `py: Python<'_>`, used `cursor.take()` to move cursor
  into `py.detach()` closure (Rust ownership requirement), performed read without
  GIL, restored cursor after completion.
- `PyDaftFile::seek`: same take/restore pattern as `read()`.
- `PyDaftFile::_supports_range_requests`: added `py: Python<'_>`, wrapped
  `block_within_async_context` call in `py.detach()` with cloned `uri` and `source`.

All changes follow the existing `py.detach()` pattern used by `exists()` at
line 102 in the same file. No behavioral changes — same inputs, same outputs,
only the GIL hold timing differs.

## Related Issues

Closes #7256

### CODE REVIEW

**Review by universalmind303 — COMMENTED** [review]
@zhouwenjiajia, this approach seems sound to me, The greptile comments seem legit though. Once those are addressed, i think we can merge.

File: src/daft-file/src/python.rs [hunk 1]
```diff
@@ -147,52 +148,61 @@ impl PyDaftFile {
...
+
+        match result {
+            Ok((buffer, bytes_read, is_read_all, cursor_back)) => {
+                self.inner.cursor = Some(cursor_back);
+                if is_read_all {
+                    self.inner.position = bytes_read;
+                } else {
+                    self.inner.position += bytes_read;
+                }
+                Ok(buffer)
```
[the remark this answers is not in the document]
Comment (zhouwenjiajia) [reply src/daft-file/src/python.rs:207]: fixed

File: src/daft-file/src/python.rs [hunk 2]
```diff
@@ -204,17 +214,27 @@ impl PyDaftFile {
...
+                .map_err(|e| PyIOError::new_err(e.to_string()))?;
+            Ok((new_pos, cursor))
+        });
+
+        match result {
+            Ok((new_pos, cursor_back)) => {
+                self.inner.cursor = Some(cursor_back);
+                self.inner.position = new_pos as usize;
+                Ok(new_pos)
+            }
```
[the remark this answers is not in the document]
Comment (zhouwenjiajia) [reply src/daft-file/src/python.rs:247]: Already fixed — error paths now carry the cursor back via the closure return value and restore it in the outer Err branch, so a failed seek/read no longer drops the open handle.

**Comment by zhouwenjiajia** [comment]
@universalmind303 Thanks for the review! The greptile comments have been addressed — both read() and seek() now return the cursor in error paths, so a failed I/O no longer drops the open handle. All CI checks are passing. Ready for another look when you have a moment.

**Review by universalmind303 — APPROVED** [review]
thanks @zhouwenjiajia
